### ISSUE TITLE
Better error reporting

### ISSUE THREAD

**Issue by antoyo** [antoyo/relm#27]
That would require the ability to set positions (spans) in the proc macro and the ability to spans errors.

**Comment by antoyo** [comment]
To have that right now, I could implement something that will run when an environment variable is set. It will do the following:
 * Try to compile the source code.
 * If there's an error within a `#[widget]` attribute:
   * Generate the source code annotated by this attribute and compile it.
   * Match the errors of the generated code with the original one.
   * Show these errors instead of the one produced by the compiler.

That will be slow, but still faster that doing that by hand.

**Comment by Geobert** [comment]
This is really needed, especially for noobs like me in Rust and you get this kind of error:
```
error: borrowed value does not live long enough
  --> src\widgets\daywidget.rs:34:9
   |
34 | #[widget]
   | --------^
   | |       |
   | |       temporary value dropped here while still borrowed
   | temporary value created here
   |
   = note: values in a scope are dropped in the opposite order they are created
```

I think I'll give up on relm for the moment :(

**Comment by antoyo** [comment]
I think that I'll soon be able to fix this issue (at least on nightly) since `syn` [was fixed](https://github.com/dtolnay/syn/pull/146).
However, I think I need this [pull request](https://github.com/rust-lang/rust/pull/40939) to be merged, but hopefully it will happen soon.

Until then, you can enable the log when you use relm by doing:
```
export RUST_LOG=warn
```
and you'll see the generated code.
Then, you can try the generated code to have a better error but it is annoying to do that.

**Comment by Geobert** [comment]
Oh cool, I'm gonna give this a shot to see if I can understand which value causes this error :)

Thank you. And to be fair, I know the whole Rust sphere is very young and still craving its way to provide all the tools needed to make a language successful :) So kudo to you for relm, it makes gtk way less ugly to use in Rust!

**Comment by antoyo** [comment]
Oh, I've just noticed that this logging feature is only in the `feature/futures-glib` branch (which contains many breaking changes).
Do you want that I add it in the master branch?

**Comment by Geobert** [comment]
If the PR needed to makes thing right by managing the current ticket don't bother, I can wait :)

**Comment by superlou** [comment]
Looks like the pull-request was merged.  Is it something we need to update on our end, or does the library need to be updated?

**Comment by antoyo** [comment]
@superlou: Unfortunately, this PR was merged with [a hack](https://github.com/rust-lang/rust/issues/43081), so it does not work right now.
Also, that will need to update the code to switch to proc-macro2.

**Comment by antoyo** [comment]
@zmitchell I talked to you about the idea of using (sy)nom, but I'm not sure we can use it for the whole thing.
I think we will only be able to parse the content of the `view!` macro with it.

**Comment by zmitchell** [comment]
I agree, I think it's the only part that it needs to be used for. I don't think I implied otherwise, did I?

**Comment by antoyo** [comment]
@zmitchell No, but I did :) .

**Comment by antoyo** [comment]
@zmitchell: I started to fix this and have some basic examples (like `buttons-attribute`) working.
You can see my work in branch [`fix/better-error`](https://github.com/antoyo/relm/tree/fix/better-error) (see [this giant commit](https://github.com/antoyo/relm/commit/466d2ba757d1cb16e158da9568f1d75fe08e6cac)).
Hopefully, I'll finish the work in 2-3 weeks.

**Comment by antoyo** [comment]
Fixed by PR #90.
